src/TodoList.tsx

```diff
@@ -0,0 +1,174 @@
+import { useState } from 'react';
+import { Todo } from './types/Todo';
+import { TodoItem } from './utils/TodoItem';
+
+interface TodoListProps {
+  todos: Todo[];
+  setRemoveTodoIsClicked: React.Dispatch<React.SetStateAction<boolean>>;
+  setEditTodoIsClicked: React.Dispatch<React.SetStateAction<boolean>>;
+  tempTodo: Todo | null;
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>;
+  setToggleTodoIsFailed: React.Dispatch<React.SetStateAction<boolean>>;
+  updatingTodosToggle: number[];
+  deletingCompletedTodos: number[];
+  setIsInputDisabled: React.Dispatch<React.SetStateAction<boolean>>;
+}
+
+export const TodoList: React.FC<TodoListProps> = ({
+  todos,
+  setRemoveTodoIsClicked,
+  setEditTodoIsClicked,
+  tempTodo,
+  setTodos,
+  setToggleTodoIsFailed,
+  updatingTodosToggle,
+  deletingCompletedTodos,
+  setIsInputDisabled,
+}) => {
+  const [itemToDelete, setItemToDelete] = useState<number | null>(null);
+  const [itemToToggle, setItemToToggle] = useState<number | null>(null);
+  const [itemToUpdate, setItemToUpdate] = useState<number | null>(null);
+  const [itemToLoad, setItemToLoad] = useState<number | null>(null);
+
+  const toggleTodo = async (todoId: number) => {
+    setItemToToggle(todoId);
+
+    const todoToUpdate = todos.find((todo) => todo.id === todoId);
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const { id, userId, title } = todoToUpdate!;
+
+    const updatedTodo: Todo = {
+      id,
+      userId,
+      title,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      completed: !todoToUpdate!.completed,
+    };
+
+    try {
+      await fetch(`https://mate.academy/students-api/todos/${todoId}`,
+        {
+          method: 'PATCH',
+          headers: {
+            'Content-Type': 'application/json;charset=utf-8',
+          },
+          body: JSON.stringify(updatedTodo),
+        });
+      // eslint-disable-next-line no-console
+      console.log('todo has been updated');
+      if (todoToUpdate) {
+        todoToUpdate.completed = !todoToUpdate.completed;
+        setTodos([...todos]);
+      }
+    } catch {
+      setToggleTodoIsFailed(true);
```

Comment:
You should handle specific error types in the catch block. This will allow you to handle different types of errors differently and provide more descriptive error messages.